agent/reporter.go

```diff
@@ -102,27 +124,30 @@ func (sr *statusReporter) run(ctx context.Context) {
 			return
 		}
 
-		for taskID, status := range sr.statuses {
-			delete(sr.statuses, taskID) // delete the entry, while trying to send.
-
-			sr.mu.Unlock()
-			err := sr.reporter.UpdateTaskStatus(ctx, taskID, status)
-			sr.mu.Lock()
-
-			// reporter might be closed during UpdateTaskStatus call
-			if sr.closed {
-				return
-			}
-
-			if err != nil {
-				log.G(ctx).WithError(err).Error("status reporter failed to report status to agent")
-
-				// place it back in the map, if not there, allowing us to pick
-				// the value if a new one came in when we were sending the last
-				// update.
-				if _, ok := sr.statuses[taskID]; !ok {
-					sr.statuses[taskID] = status
-				}
+		// swap out the statuses map, so we can release the lock while we
+		// do the batch
+		statuses := sr.statuses
+		sr.statuses = map[string]*api.TaskStatus{}
+		// unlock the map so new statuses can be added while we process the
+		// current ones
+		sr.mu.Unlock()
+		err := sr.reporter.UpdateTaskStatus(ctx, statuses)
+		// re-lock the map so we can add statuses back if we need to.
+		sr.mu.Lock()
+		// it's possible that the status reporter may have closed while the
+		// lock was released, but as the code is structured now, that's
+		// unimportant. we might add some statuses back to the map, which would
+		// be useless, but when we back get around to the sr.closed check
+		// above, we'll exit anyway.
+		if err != nil {
```

Comment:
The granularity is going to problematic here: there are lots of cases where a single task status could result in an error, causing the group to get retried forever. For example, "not found" for a task status update should result in dropping the status, but that would result in a retry of the group.

Reply:
i recognize this. there's a problem of granularity in the opposite direction too, though, where if multiple statuses are bad, we keep having to retry.

Reply:
i'd like your rough take on solution.

Reply:
Oh, FWIW, the only error that can result from a bad task is if a node tries to send an update for a task it is not assigned. We can fossilize this in the API if we'd like.